### DISCUSSION TITLE
setBuffers useage question

### DISCUSSION THREAD

**Discussion by jimtal001** [earlephilhower/arduino-pico#2018] [General]
Hello Earle,  Thank you for your great work here.

Can you explain the use of the setBuffers function used for I2S Input?  I am currently working with a single microphone (mono).  I've searched for an example or documentation which explains how to use this function parameters but have thus far been unsuccessful in finding one.  Maybe you could share an example?

**setBuffers(size_t buffers, size_t bufferWords, int32_t silenceSample = 0);**

Thanks!

**Comment by earlephilhower** [comment]
That just sets the amount of RAM that's allocated for buffering input streams.  The way I2S works is it fills up `bufferWords` in one buffer then notifies the app that data is available.  So, `bufferWords` sets the latency before your app is notified, so too large is not great for real-time apps, but too-small will cause excessive IRQs and may overflow.

`buffers` is the total number of RAM buffers to allocate, each of size `bufferWords`.  You need at least 3 or I2S cannot safely function (ping, pong, and app owned) but there's no particular harm in having more.

Rough guesstimate, I'd use something like 32 or 64 for `bufferWords` and `4 or 5 for `buffers`.

**Comment by jimtal001** [comment]
Thanks for your help!

Is there a way to call a read command to load the I2S buffer without using interrupts? 
Like the esp32 sample code below:

`  
void loop() {
  size_t bytesIn = 0;
  esp_err_t result = i2s_read(...);  // load I2S data and place in data buffer

  if (result == ESP_OK)
  {
    // Read I2S data buffer
    int16_t samples_read = bytesIn / 8;
    if (samples_read > 0) {
        // ... do something
      }
    }
}`

**Reply by earlephilhower** [reply]
Sure, you don't have to use the `onReceive` callback and can just use something like
````
while (audio.available()) {
    int16_t l, r;
    audio.read16(&l, &r);
    /* process... */
}
````

See the example [here](https://github.com/earlephilhower/arduino-pico/blob/master/libraries/I2S/examples/I2SInput/I2SInput.ino).

**Comment by jimtal001** [comment]
Does the bool available() wait for the buffer to fill before becoming
true?  The example you linked doesn't set the I2S buffer.

**Reply by earlephilhower** [reply]
`available()` does what `available()` does in a `Stream`, lets you know if you can read data w/o blocking.  So it'll return `false` until there's some data transferred from the HW to the app side.

**Comment by jimtal001** [comment]
Would you please explain the **Boolean** returned by an I2S read statement like: **read16(...)**. 
Does **true** mean data was returned? Or something else, like time-out, I2S-error, etc.

As per the I2S.h file:  "Read samples from I2S port, will **block** until data available", so this would indicate that the returned Boolean does not mean data available.

**Reply by earlephilhower** [reply]
They do block, and always return `true` when data comes back.  It can only return `false` if the `I2S` wasn't started or is OUTPUT only.
